banyan/src/forest/read.rs

```diff
@@ -19,7 +13,291 @@ use anyhow::{anyhow, Result};
+    forest: Forest<T, V, R>,
+    offset: u64,
+    query: Q,
+    mk_extra: F,
+    stack: SmallVec<[TraverseState<T>; 5]>,
+    mode: Mode,
+}
+
+struct TraverseState<T: TreeTypes> {
+    index: Arc<Index<T>>,
+    // If `index` points to a branch node, `position` points to the currently
+    // traversed child
+    position: usize,
+    // For each child, indicates whether it should be visited or not. This is
+    // initially empty, and initialized whenver we hit a branch.
+    // Branches can not have zero children, so when this is empty we know that we have
+    // to initialize it.
+    filter: SmallVec<[bool; 64]>,
+}
+
+impl<T: TreeTypes> TraverseState<T> {
+    fn new(index: Arc<Index<T>>, mode: &Mode) -> Self {
+        let position = match mode {
+            Mode::Forward => 0,
+            // We don't know yet the number of DIRECT children this index has,
+            // so this needs to be initialized after resolving a potential
+            // branch
+            Mode::Backward => usize::MAX,
+        };
+        Self {
+            index,
+            position,
+            filter: smallvec![],
+        }
+    }
+    fn is_exhausted(&self, mode: &Mode) -> bool {
+        match mode {
+            Mode::Forward => !self.filter.is_empty() && self.position == self.filter.len(),
+            Mode::Backward => self.position == 0,
+        }
+    }
+    fn next_pos(&mut self, mode: &Mode) {
+        match mode {
+            Mode::Forward => self.position += 1,
+            Mode::Backward => self.position -= 1,
+        }
+    }
+}
+
+impl<T: TreeTypes, V, R, Q, E, F> ForestIter<T, V, R, Q, F>
+where
+    T: TreeTypes + 'static,
+    V: DagCbor + Clone + Send + Sync + Debug + 'static,
+    R: ReadOnlyStore<T::Link> + Clone + Send + Sync + 'static,
+    Q: Query<T> + Clone + Send + 'static,
+    E: Send + 'static,
+    F: Fn(IndexRef<T>) -> E + Send + Sync + 'static,
+{
+    pub(crate) fn new(
+        forest: Forest<T, V, R>,
+        query: Q,
+        index: Arc<Index<T>>,
+        mk_extra: F,
+    ) -> Self {
+        let mode = Mode::Forward;
+        let stack = smallvec![TraverseState::new(index, &mode)];
+
+        Self {
+            forest,
+            offset: 0,
+            query,
+            mk_extra,
+            stack,
+            mode,
+        }
+    }
+    pub(crate) fn new_rev(
+        forest: Forest<T, V, R>,
+        query: Q,
+        index: Arc<Index<T>>,
+        mk_extra: F,
+    ) -> Self {
+        let offset = index.count();
+        let mode = Mode::Backward;
+        let stack = smallvec![TraverseState::new(index, &mode)];
+
+        Self {
+            forest,
+            offset,
+            query,
+            mk_extra,
+            stack,
+            mode,
+        }
+    }
+}
+
+// This is not exposed, as `ForestIter` has to be constructed with the proper
+// initial values for reverse iteration
+impl<T: TreeTypes, V, R, Q, E, F> DoubleEndedIterator for ForestIter<T, V, R, Q, F>
```

Comment:
Does this make sense to implement? Basically a proper double ended iterator can be iterated from both sides, even interleaved, but ours can only be forward all the time or backward all the time.

Reply:
No, it doesn't. With the two constructors for ForestIter, we can get away with only one `impl Iterator`.